internal/pkgs/wifimanager/controller.go

```diff
@@ -4,101 +4,444 @@ package wifimanager
+		}
+
+		// Register StateChanged signal
+		channel := make(chan *dbus.Signal, 1)
+		c.conn.Signal(channel)
+		err = c.conn.AddMatchSignal(
+			dbus.WithMatchInterface("org.freedesktop.NetworkManager.Connection.Active"),
+			dbus.WithMatchMember("StateChanged"),
+			dbus.WithMatchObjectPath(activeConnectionPath),
+		)
+		if err != nil {
+			return "", fmt.Errorf("failed to add match signal: %s", err)
+		}
+
+		// Check the connection activated successfully
+		for signal := range channel {
+			fmt.Println(signal.Name)
+			if signal.Name != "org.freedesktop.NetworkManager.Connection.Active.StateChanged" {
+				continue
+			}
+			switch state := signal.Body[0].(uint32); state {
+			case NmActiveConnectionStateUnknown:
+				return "", fmt.Errorf("unexpected active connection state: unknown")
+			case NmActiveConnectionStateActivating:
+				response = "Connecting to " + SSID
+			case NmActiveConnectionStateActivated:
+				response = "Connected to " + SSID + " successfully"
+				return response, nil
+			case NmActiveConnectionStateDeactivating:
+				return "", fmt.Errorf("unexpected active connection state: deactivating")
+			case NmActiveConnectionStateDeactivated:
+				return "", fmt.Errorf("unexpected active connection state: deactivated")
+			default:
+				return "", fmt.Errorf("unknown active connection state")
+			}
+		}
+	}
+	return string(response), nil
+}
+
+func (c *WifiController) Disconnect(ctx context.Context) (string, error) {
+	response := "wifi disconnection is failed"
+
+	// Input validation
+	if ctx == nil {
+		return "", fmt.Errorf("context cannot be nil")
 	}
 
+	// Iterate over Wi-Fi devices
+	for _, device := range c.wifiDevices {
+		// Add a new connection and connect
+		disconnect := device.Call("org.freedesktop.NetworkManager.Device.Disconnect", 0)
+		if disconnect.Err != nil {
+			return "", fmt.Errorf("failed to disconnect %s: %s", device.Path(), disconnect.Err)
+		} else {
+			var intf string
+			interfaceVariant, _ := device.GetProperty("org.freedesktop.NetworkManager.Device.Interface")
+			_ = interfaceVariant.Store(&intf)
+			response = intf + " disconnected successfully"
+		}
+	}
 	return string(response), nil
 }
 
 func (c *WifiController) WifiRadioSwitch(ctx context.Context, TurnOn bool) (string, error) {
+	var wifiEnabled bool
+	var status string
 
 	// Input validation
 	if ctx == nil {
 		return "", fmt.Errorf("context cannot be nil")
 	}
 
-	nmcliRunCmd := "/run/current-system/sw/bin/nmcli"
-	nmcliRunCmd += " radio wifi "
+	// Check the wifi status
+	wifiEnabledVariant, err := c.nm.GetProperty("org.freedesktop.NetworkManager.WirelessEnabled")
+	if err != nil {
+		return "", fmt.Errorf("failed to get wifi status: %s", err)
+	}
+	err = wifiEnabledVariant.Store(&wifiEnabled)
+	if err != nil {
+		return "", fmt.Errorf("failed to get wifi status: %s", err)
+	}
 
 	if TurnOn {
-		nmcliRunCmd += "on"
+		status = "enabled"
 	} else {
-		nmcliRunCmd += "off"
+		status = "disabled"
 	}
```

Comment:
This seems to be missing a call to `SetProperty`.

Reply:
Thanks to point that. It is removed unintentionally when refactoring. Added back.